File: Typecast/TCResourceReference.m
//
//  TCResourceReference.m
//  Type Designer
//
//  Created by David Schweinsberg on 28/12/12.
//  Copyright (c) 2012 David Schweinsberg. All rights reserved.
//

#import "TCResourceReference.h"
#import "TCDataInput.h"

@implementation TCResourceReference

- (id)initWithDataInput:(TCDataInput *)dataInput
{
    self = [super init];
    if (self)
    {
        _resourceID = [dataInput readUnsignedShort];
        _nameOffset = [dataInput readShort];
        _attributes = [dataInput readUnsignedByte];
        _dataOffset = ([dataInput readUnsignedByte] << 16) | [dataInput readUnsignedShort];
        _handle = [dataInput readInt];
    }
    return self;
}

- (void)readName:(TCDataInput *)dataInput
{
    if (_nameOffset > -1)
    {
        int len = [dataInput readUnsignedByte];
        NSData *data = [dataInput readDataWithLength:len];
        _name = [[NSString alloc] initWithData:data encoding:NSASCIIStringEncoding];
    }
}

@end

File: Typecast/TCResourceReference.swift
//
//  TCResourceReference.swift
//  Type Designer
//
//  Created by David Schweinsberg on 7/27/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Foundation

class TCResourceReference {
  let resourceID: Int
  let nameOffset: Int
  let attributes: Int
  let dataOffset: Int
  let handle: Int
  var name: String?

  init(dataInput: TCDataInput) {
    resourceID = Int(dataInput.readUnsignedShort())
    nameOffset = Int(dataInput.readShort())
    attributes = Int(dataInput.readUnsignedByte())
    dataOffset = Int(dataInput.readUnsignedByte()) << 16 | Int(dataInput.readUnsignedShort())
    handle = Int(dataInput.readInt())
  }

  func readName(dataInput: TCDataInput) {
    if nameOffset > -1 {
      let len = UInt(dataInput.readUnsignedByte())
      let data = dataInput.readData(withLength: len)
      name = String(data: data!, encoding: String.Encoding.ascii)!
    }
  }
}
